src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rest/QueryBatchRequestHandler.java

```diff
@@ -0,0 +1,309 @@
+      startTime -= startTime % samplingPeriod;
+      endTime -= endTime % samplingPeriod;
+      if (startTime == endTime) {
+        throw new InvalidParameterException("starttime and endtime cannot be equal when rounded down to the nearest sampling period");
+      }
+      if (endTime > currentTime) {
+        throw new InvalidParameterException("endtime can be no greater than the system time at the node");
+      }
+      if (startTime < currentTime - PluginSettings.instance().getBatchMetricsRetentionPeriodMinutes() * 60 * 1000) {
+        throw new InvalidParameterException("starttime must be within the retention period");
+      }
+
+      String queryResponse = queryFromBatchMetrics(batchMetrics, metrics, startTime, endTime, samplingPeriod,
+              DEFAULT_MAX_DATAPOINTS);
+      sendResponse(exchange, queryResponse, HttpURLConnection.HTTP_OK);
+    } catch (InvalidParameterException e) {
+      LOG.error(
+              (Supplier<?>)
+                      () ->
+                              new ParameterizedMessage(
+                                      "QueryException {} ExceptionCode: {}.",
+                                      e.toString(),
+                                      StatExceptionCode.REQUEST_ERROR.toString()),
+              e);
+      StatsCollector.instance().logException(StatExceptionCode.REQUEST_ERROR);
+      String response = "{\"error\":\"" + e.getMessage() + ".\"}";
+      sendResponse(exchange, response, HttpURLConnection.HTTP_BAD_REQUEST);
+    } catch (Exception e) {
+      LOG.error(
+              (Supplier<?>)
+                      () ->
+                              new ParameterizedMessage(
+                                      "QueryException {} ExceptionCode: {}.",
+                                      e.toString(),
+                                      StatExceptionCode.REQUEST_ERROR.toString()),
+              e);
+      StatsCollector.instance().logException(StatExceptionCode.REQUEST_ERROR);
+      String response = "{\"error\":\"" + e.toString() + "\"}";
+      sendResponse(exchange, response, HttpURLConnection.HTTP_INTERNAL_ERROR);
+    }
+  }
+
+  @VisibleForTesting
+  public int appendMetrics(Long timestamp, List<String> metrics, StringBuilder builder, int maxDatapoints) throws Exception {
+    maxDatapoints += 1;
+    builder.append("\"");
+    builder.append(timestamp);
+    builder.append("\":{");
+    MetricsDB db = MetricsDB.fetchExisting(timestamp);
+    for (int metricIndex = 0, numMetrics = metrics.size(); metricIndex < numMetrics; metricIndex++) {
+      String metric = metrics.get(metricIndex);
+      Result<Record> results = db.queryMetric(metric, MetricsModel.ALL_METRICS.get(metric).dimensionNames, maxDatapoints);
+      if (results != null) {
+        maxDatapoints -= results.size();
+        if (maxDatapoints == 0) {
+          throw new InvalidParameterException(String.format("requested data exceeds the %d datapoints limit", DEFAULT_MAX_DATAPOINTS));
+        }
```

Comment:
`results.size() cannot be ` cannot be > `maxDatapoints`, since `maxDatapoints` is supplied as the limit when invoking `db.queryMetric`.

`DEFAULT_MAX_DATAPOINTS` cannot be overridden. It is an internal safeguard to ensure that the user does not break their system by exceeding a memory limit. Even if we allowed the user to configure the max datapoints, we would still need an internal limit to ensure no excessive values are supplied. Until more clear internal limits are identified, we ought to leave configurability off the table in favor of safety. Additionally, these changes may further be obviated if pagination is introduced in the future.

Reply:
What I meant was that `if (maxDatapoints == 0)` is a very exact check. We might want `<=` instead?

Reply:
Logically, maxDatapoints should never be less than zero. But, sure.